### PR TITLE
Feature 25. Like feedback. narek

### PR DESCRIPTION
**Pull request by narekacy**
One kindly request for reviewers:
- when like count is zero, there is no 'zero' shown for user. But **when feedback is liked**, user sees '1' in total likes count. **This moves 'dislike' button to right side**. It is needed to fix its position. Do somebody know how to implement this?

Also little bug was fixed in this task connected with feedback regex we don't use anymore.

**EDITED 12.05.2020:**
@vitalii-mak **[suggested](https://github.com/Social-projects-Rivne/Lunch_Time/pull/106#pullrequestreview-408043099)** saving our DB architecture while doing our tasks. I agree. Today I had to implemet this and I got stuck when passing likes/dislikes count via `FeedbackMarkDto`. To be clear, we pass from client part `FeedbackMarkDto` RequestBody with `Long feedbackId` and `Long personId`. But when we pass `FeedbackMarkDto` to client, we need `Set<Long> personIdLikes` and `Set<Long> personIdDislikes` to match their quantity on client part. This brings to trouble when planning **correct** app architecture. I had similar questions as, what to use, `FeedbackDto` or `FeedbackMarkDto`, do we need `FeedbackMarkController` or `FeedbackMarkRepository`? Due to my poor experience I stopped doing my task in unknown way and did it as it works right now.

I'm honestly interested in better and flexible way to build app's architecture. But, due to no additional time for our final demo and holiday it was difficult to discuss this with our team and mentor. So I left this to current working condition.

### CODE REVIEW

File: client/src/components/feedback/comment.js [hunk 1]
```diff
@@ -29,18 +35,61 @@ class FeedbackComment extends Component {
...
+              item.likes = response.data.likes;
+              return { item };
+            });
+          }
+        } else {
+          setTimeout(() => {
+            this.setState({
+              unexpectedError: false,
+            });
+          }, 3000);
```
Comment (vitaliimak) [inline client/src/components/feedback/comment.js:82]: Could you remove all timeouts?
In 2020 we have very fast devices and all developers and services and companies try to make applications faster but you add additional complications in code and UX. Very bad idea

File: client/src/components/feedback/comment.js [hunk 2]
```diff
@@ -1,15 +1,21 @@
 import React, { Component } from 'react';
-import { Container, Image } from 'react-bootstrap';
+import {
+  Container, Image, Badge,
+} from 'react-bootstrap';
 import '../../styles/feedback-comment.css';
 import PropTypes from 'prop-types';
 import Api from '../../services/api';
+import Person from '../../services/person';
```
Comment (vitaliimak) [inline client/src/components/feedback/comment.js]: Don't use this class. I've removed it in https://github.com/Social-projects-Rivne/Lunch_Time/pull/109
You can write similar code, as in my PR, or wait to merge it in dev and update your code

**Review by vitaliimak — CHANGES_REQUESTED** [review]
image

File: client/src/components/feedback/comment.js (hunk 1 above)
Comment (narekacy) [reply client/src/components/feedback/comment.js:82]: This timeout is for notification show. We can't show static notification. In some period after show this should be hidden. That is what timeout does here.

File: client/src/components/feedback/comment.js (hunk 2 above)
Comment (narekacy) [reply client/src/components/feedback/comment.js]: Ok, let your PR be merged first. Then I will merge with dev.

File: client/src/components/feedback/comment.js (hunk 1 above)
Comment (vitaliimak) [reply client/src/components/feedback/comment.js:82]: React bootstrap has highlevel handlers for similar cases https://react-bootstrap.github.io/components/toasts/#autohide
https://react-bootstrap.github.io/components/alerts/#api - onClose() method
You should use alert or toast for messages, not bages

Comment (narekacy) [reply client/src/components/feedback/comment.js:82]: I have discovered information by this links. So, I want to say that due to my current implementaton badge (or text if we refuse using badge) is shown immediately near like/unlike button and doesn't change whole feedback card views. Both toast and alert will push the rest part of page to be situated when shown. 

I see multiple desicions:
1. We don't use timeout, but then we can show notification till user moves to another page.
2. Don't use any notification, but hover message as it is right now.

All in all, I believe there is a better way to implement this, even with react component lifecycle methods - I mean stop rendering component when it is inactive. But, due to the fact we are not UI developers, I'd leave it like it is right now till our memory leak issue can appear as potential bug.

File: client/src/components/feedback/comment.js [hunk 3]
```diff
@@ -49,14 +98,51 @@ class FeedbackComment extends Component {
...
+              <img
+                src={likeImage}
+                width="25px"
+                height="25px"
+                alt="like"
+                title={title}
+                style={{ cursor: 'pointer' }}
+                onClick={onClick}
+              />
+              <small>{item.likes.length > 0 ? item.likes.length : ''}</small>
```
Comment (vitaliimak) [inline client/src/components/feedback/comment.js:141]: > when like count is zero, there is no 'zero' shown for user. But when feedback is liked, user sees '1' in total likes count. This moves 'dislike' button to right side. It is needed to fix its position. Do somebody know how to implement this?

You can use 0 instead `''`

**Review by vitaliimak — CHANGES_REQUESTED** [review]
I think you should do this task https://trello.com/c/AuVoxl8W/27-26as-an-authorized-user-of-the-system-i-want-to-have-a-possibility-to-dislike-feedback too in on iteration

Why don't you follow our database scheem?
image

**Comment by narekacy** [comment]
@vitalii-mak @Andrii35 @Melnyk9506 

PR is updated

**Comment by narekacy** [comment]
Could someone try to merge locally my branch with dev and tell what is result? When I do this and resolve minor conflict, my client part isn't working. Tried to start it again, nothing helped.

image

**Comment by narekacy** [comment]
> Why don't you follow our database scheem?

I was inattentive here. Will fix it now.

File: client/src/components/feedback/comment.js (hunk 3 above)
Comment (narekacy) [reply client/src/components/feedback/comment.js:141]: Ok, but due to top web-apps, it is not good experience to show '0' value for user

Comment (narekacy) [reply client/src/components/feedback/comment.js:141]: I'm gonna do this task after this one is done successfully. https://trello.com/c/AuVoxl8W/27-26as-an-authorized-user-of-the-system-i-want-to-have-a-possibility-to-dislike-feedback

Finally, there are two tasks in trello

**Review by Andrii35 — COMMENTED** [review]
I have an Error by start of the server:

lunch-time-server | [ERROR] COMPILATION ERROR : 
lunch-time-server | [INFO] -------------------------------------------------------------
lunch-time-server | [ERROR] /app/src/main/java/com/lunchtime/mapper/FeedbackMapper.java:[55,17] method map in interface java.util.stream.Stream<T> cannot be applied to given types;
lunch-time-server |   required: java.util.function.Function<? super com.lunchtime.models.Like,? extends R>
lunch-time-server |   found: Person::getId
lunch-time-server |   reason: cannot infer type-variable(s) R
lunch-time-server |     (argument mismatch; invalid method reference
lunch-time-server |       method getId in class com.lunchtime.models.Person cannot be applied to given types
lunch-time-server |         required: no arguments
lunch-time-server |         found: com.lunchtime.models.Like
lunch-time-server |         reason: actual and formal argument lists differ in length)
lunch-time-server | [ERROR] /app/src/main/java/com/lunchtime/mapper/FeedbackMapper.java:[55,22] invalid method reference
lunch-time-server |   non-static method getId() cannot be referenced from a static context
lunch-time-server | [ERROR] /app/src/main/java/com/lunchtime/controllers/FeedbackController.java:[43,38] method likeFeedback in interface com.lunchtime.service.FeedbackService cannot be applied to given types;
lunch-time-server |   required: java.lang.Long,java.lang.Long
lunch-time-server |   found: java.lang.Long,java.lang.Long,java.lang.Boolean
lunch-time-server |   reason: actual and formal argument lists differ in length
lunch-time-server | [ERROR] /app/src/main/java/com/lunchtime/service/impl/FeedbackServiceImpl.java:[47,32] no suitable method found for add(com.lunchtime.models.Person)
lunch-time-server |     method java.util.Collection.add(com.lunchtime.models.Like) is not applicable
lunch-time-server |       (argument mismatch; com.lunchtime.models.Person cannot be converted to com.lunchtime.models.Like)
lunch-time-server |     method java.util.Set.add(com.lunchtime.models.Like) is not applicable
lunch-time-server |       (argument mismatch; com.lunchtime.models.Person cannot be converted to com.lunchtime.models.Like)
lunch-time-server | [INFO] 4 errors 
lunch-time-server | [INFO] -------------------------------------------------------------
lunch-time-server | [INFO] ------------------------------------------------------------------------
lunch-time-server | [INFO] BUILD FAILURE
lunch-time-server | [INFO] ------------------------------------------------------------------------
lunch-time-server | [INFO] Total time:  14.182 s
lunch-time-server | [INFO] Finished at: 2020-05-09T17:56:48Z

**Comment by narekacy** [comment]
> I have an Error by start of the server:
> 
> lunch-time-server | [ERROR] COMPILATION ERROR :
> ...
> lunch-time-server | [INFO] Finished at: 2020-05-09T17:56:48Z

Fixed. Check if it is ok, please. But first, we need to decide how to store our likes. Check our conversetion in **[#development](https://rv-054java.slack.com/archives/CUKLK9279)**

**Comment by vitaliimak** [comment]
@narekacy is this PR ready to review?

**Comment by narekacy** [comment]
> @narekacy is this PR ready to review?

No, I want to move it to 'draft' (in progress), but haven't received answer how to do that yet. May be someone knows how to move this PR to 'in progress' status?

**Comment by vitaliimak** [comment]
@narekacy 
image

**Review by vitaliimak — APPROVED** [review]
I can add +1 to my own feedback)
image

**Comment by narekacy** [comment]
> I can add +1 to my own feedback)
> image

Yes) User shouldn't have this possibility?

**Comment by vitaliimak** [comment]
> Yes) User shouldn't have this possibility?

I think no

**Comment by narekacy** [comment]
> > Yes) User shouldn't have this possibility?
> 
> I think no

This is limiting users features in our app. If we discover top apps experience, no one restricts user to like (or dislike if it is present) his own post.

**Review by Andrii35 — APPROVED** [review]

**Review by Melnyk9506 — APPROVED** [review]

**Pull request merged**